File: src/gethgl.f
C Output from Public domain Ratfor, version 1.03
      subroutine gethgl(nd,ndistr,fy,gmu,sd,lambda,p,ashp,bshp,phimat,y,
     *ymiss, tdm,nind,n,tpm,xispd,size,nbot,ntop,d1pi,d2pi,kstate, npar,
     *npt,nyv,nxc,d1p,d2p,d1f,d2f,d1a,d1b,d2aa,d2ab,d2bb, alpha,alphw,a,
     *b,aw,bw,xlc,hess)
      implicit double precision(a-h,o-z)
      integer ymiss(n), size
      dimension fy(kstate,n),gmu(kstate,n),sd(kstate,n)
      double precision lambda(kstate,n)
      dimension p(kstate,n)
      dimension ashp(kstate,n),bshp(kstate,n),phimat(nxc,nyv)
      dimension y(n),tdm(nxc,nind)
      dimension tpm(kstate,kstate),xispd(kstate)
      dimension d1pi(kstate,npt),d2pi(kstate,npt,npt)
      dimension d1p(kstate,kstate,npt),d2p(kstate,kstate,npt,npt)
      dimension d1f(kstate,npar),d2f(kstate,npar,npar)
      dimension d1a(kstate),d1b(kstate)
      dimension d2aa(kstate),d2ab(kstate),d2bb(kstate)
      dimension alpha(kstate),alphw(kstate)
      dimension a(kstate,npar),b(kstate,npar,npar)
      dimension aw(kstate,npar),bw(kstate,npar,npar)
      dimension xlc(n),hess(npar,npar)
      zero = 0.d0
      kt = 1
      if(nd .ge. 1)then
      kstart = 1
      call derivf(ndistr,y(kt),ymiss(kt),fy(1,kt),phimat,tdm(1,kstart), 
     *gmu(1,kt),sd(1,kt),lambda(1,kt),p(1,kt),ashp(1,kt), bshp(1,kt),kst
     *ate,npar,npt,nyv,nxc,size,nbot,ntop,d1a,d1b, d2aa,d2ab,d2bb,nd,d1f
     *,d2f)
      endif
      sxlc = zero
      kstpr = npt - npar
      do23002 j = 1,kstate 
      alpha(j) = xispd(j)*fy(j,kt)
      sxlc = sxlc + alpha(j)
      if(nd .ge. 1)then
      do23006 k1 = 1,npar 
      if(ymiss(1) .eq. 1)then
      d1fx1 = zero
      else
      d1fx1 = d1f(j,k1)
      endif
      a(j,k1) = xispd(j)*d1fx1 + fy(j,kt)*d1pi(j,kstpr+k1)
      if(nd .eq. 2)then
      do23012 k2 = 1,npar 
      if(ymiss(1) .eq. 1)then
      d1fx2 = zero
      else
      d1fx2 = d1f(j,k2)
      endif
      if(ymiss(1) .eq. 1)then
      d2fx = zero
      else
      d2fx = d2f(j,k1,k2)
      endif
      b(j,k1,k2) = xispd(j)*d2fx + d1pi(j,kstpr+k1)*d1fx2 + d1pi(j,kstpr
     *+k2)*d1fx1 + fy(j,kt)*d2pi(j,kstpr+k1,kstpr+k2)
23012 continue
23013 continue
      endif
23006 continue
23007 continue
      endif
23002 continue
23003 continue
      xlc(1) = sxlc
      do23018 j = 1,kstate 
      alpha(j) = alpha(j)/sxlc
23018 continue
23019 continue
      if(n.gt.1)then
      do23022 kt = 2,n 
      if(nd .ge. 1)then
      kstart = 1+(kt-1)*kstate
      call derivf(ndistr,y(kt),ymiss(kt),fy(1,kt),phimat,tdm(1,kstart), 
     *gmu(1,kt),sd(1,kt),lambda(1,kt),p(1,kt),ashp(1,kt), bshp(1,kt),kst
     *ate,npar,npt,nyv,nxc,size,nbot,ntop,d1a,d1b, d2aa,d2ab,d2bb,nd,d1f
     *,d2f)
      endif
      if(nd .eq. 2)then
      do23028 j = 1,kstate 
      do23030 k1 = 1,npar 
      if(ymiss(kt) .eq. 1)then
      d1fx1 = zero
      else
      d1fx1 = d1f(j,k1)
      endif
      do23034 k2 = 1,npar 
      if(ymiss(kt) .eq. 1)then
      d1fx2 = zero
      d2fx = zero
      else
      d1fx2 = d1f(j,k2)
      d2fx = d2f(j,k1,k2)
      endif
      vvv = zero
      xxx = zero
      yy1 = zero
      yy2 = zero
      zz1 = zero
      zz2 = zero
      www = zero
      do23038 i = 1,kstate 
      vvv = vvv + alpha(i)*d2p(i,j,kstpr+k1,kstpr+k2)
      xxx = (xxx + a(i,k1)*d1p(i,j,kstpr+k2) + a(i,k2)*d1p(i,j,kstpr+k1)
     * + b(i,k1,k2)*tpm(i,j))
      yy1 = yy1 + alpha(i)*d1p(i,j,kstpr+k2)
      yy2 = yy2 + a(i,k2)*tpm(i,j)
      zz1 = zz1 + alpha(i)*d1p(i,j,kstpr+k1)
      zz2 = zz2 + a(i,k1)*tpm(i,j)
      www = www + alpha(i)*tpm(i,j)
23038 continue
23039 continue
      vvv = fy(j,kt)*vvv
      xxx = fy(j,kt)*xxx/sxlc
      yyy = d1fx1*(yy1 + yy2/sxlc)
      zzz = d1fx2*(zz1 + zz2/sxlc)
      www = d2fx*www
      bw(j,k1,k2) = vvv + xxx + yyy + zzz + www
23034 continue
23035 continue
23030 continue
23031 continue
23028 continue
23029 continue
      do23040 j = 1,kstate 
      do23042 k1 = 1,npar 
      do23044 k2 = 1,npar 
      b(j,k1,k2) = bw(j,k1,k2)
23044 continue
23045 continue
23042 continue
23043 continue
23040 continue
23041 continue
      endif
      if(nd .ge. 1)then
      do23048 j = 1,kstate 
      do23050 k = 1,npar 
      if(ymiss(kt) .eq. 1)then
      d1fx = zero
      else
      d1fx = d1f(j,k)
      endif
      xxx = zero
      yyy = zero
      zzz = zero
      do23054 i = 1, kstate 
      xxx = xxx + alpha(i)*d1p(i,j,kstpr+k)
      yyy = yyy + a(i,k)*tpm(i,j)
      zzz = zzz + alpha(i)*tpm(i,j)
23054 continue
23055 continue
      aw(j,k) = fy(j,kt)*(xxx + yyy/sxlc) + d1fx*zzz
23050 continue
23051 continue
23048 continue
23049 continue
      do23056 j = 1,kstate 
      do23058 k = 1,npar 
      a(j,k) = aw(j,k)
23058 continue
23059 continue
23056 continue
23057 continue
      endif
      sxlc = zero
      do23060 j = 1,kstate 
      alphw(j) = zero
      do23062 i = 1,kstate 
      alphw(j) = alphw(j) + alpha(i)*tpm(i,j)
23062 continue
23063 continue
      alphw(j) = fy(j,kt)*alphw(j)
      sxlc = sxlc + alphw(j)
23060 continue
23061 continue
      xlc(kt) = sxlc
      do23064 j = 1,kstate 
      alpha(j) = alphw(j)/sxlc
23064 continue
23065 continue
23022 continue
23023 continue
      endif
      if(nd .eq. 2)then
      do23068 k1 = 1,npar 
      do23070 k2 = 1,npar 
      xxx = zero
      yyy = zero
      zzz = zero
      do23072 i = 1,kstate 
      xxx = xxx + b(i,k1,k2)
      yyy = yyy + a(i,k1)
      zzz = zzz + a(i,k2)
23072 continue
23073 continue
      hess(k1,k2) = (xxx - yyy*zzz/sxlc)/sxlc
23070 continue
23071 continue
23068 continue
23069 continue
      endif
      return
      end

File: R/getHgl.R
getHgl <- function(nd,distr,theta,data,fmla,size,nbot,ntop) {
#
# Function getHgl --- get Hessian, gradient, log likelihood.
#
# Note: nd == number of derivatives to calculate.
# 0 <--> just calculate log likelihood (previously done by getLl())
# 1 <--> calculate gradient and log likelihood (previously done by getGl())
# 2 <--> calculate Hessian, gradient and log likelihood
#
dists <-c("Gaussian","Poisson","Binomial","Dbd","Multinom")
if(!(distr %in% dists)) {
    whinge <- paste0("Distribution ",distr," not recognised.\n")
    stop(whinge)
}
ndistr <- match(distr,dists)
K      <- length(levels(data[["state"]]))
if(is.null(size)) size <- 1

# Determine whether tau is to be revised/estimated in the LM step.
inclTau <- attr(theta,"inclTau")

# Dig out tau, zeta and phi from theta.
tau  <- theta$tau
zeta <- theta$zeta
phi  <- theta$phi
npt  <- length(tau) + length(zeta) + length(phi)
npar <- if(inclTau) npt else npt - length(tau) 
nms  <- c(if(inclTau) names(tau) else  NULL,names(zeta),names(phi))
#theta$tau <- reviseTau(distr,fmla,data,theta,size,nbot,ntop)

ynm  <- as.character(fmla[2])
y    <- data[[ynm]]
X    <- model.matrix(fmla[-2],data=data)
nxc  <- ncol(X)
if(distr %in% c("Gaussian","Poisson","Binomial")) {
    if(length(phi) != nxc) {
        stop("Alles upgefucken ist!\n")
    }
}

# Check that y is a factor when distr is "Multinom" (and if so,
# convert it to a numeric vector) and is numeric otherwise.  In the
# "Multinom" case, we must dig out the levels of y before doing the
# conversion to numeric mode, i.e. before buggering up y.
if(distr=="Multinom") {
    if(inherits(y,"factor")) {
        nyv   <- length(levels(y))
        y     <- as.numeric(y)
    } else {
        stop("For the Multinom distribution, \"y\" must be a factor.\n")
    }
} else {
    if(!is.numeric(y)) {
        whinge <- paste0("For the Gaussian, Poisson, Binomial and Dbd\n",
                         "  distributions, \"y\" must be numeric.\n")
        stop(whinge)
    }
    nyv <- 1
}

# Get the model components.
mc <- getModComps(distr,fmla,data,theta,size,nbot,ntop)
fy <- mc$fy

# The following components will be dummies/space-holders except when
# "distr" is the appropriate one.
gmu    <- mc$gmu    # Dummy except when distr is Gaussian.
sd     <- mc$sd     # Dummy except when distr is Gaussian.
lambda <- mc$lambda # Dummy except when distr is Poisson.
p      <- mc$p      # Dummy except when distr is Binomial.
ashp   <- mc$ashp   # Dummy except when distr is Dbd.
bshp   <- mc$bshp   # Dummy except when distr is Dbd.
Rho    <- mc$Rho    # Dummy except when distr is Multinom.


# Get tpm, and ispd
tpm  <- getTpm(theta,K)
ispd <- reviseIspd(tpm)

d12p  <- derivp(npt,K,tau,TRUE)
d1p   <- d12p$d1p
d2p   <- d12p$d2p
d12pi <- derivpi(ispd,tpm,npt,d12p)
d1pi  <- d12pi$d1pi
d2pi  <- d12pi$d2pi
nc    <- length(fy)
lcf   <- names(fy)

xll        <- numeric(nc)
names(xll) <- lcf
if(nd >= 1) {
    alist <- vector("list",nc)
    names(alist) <- lcf
}
if(nd == 2) {
    hlist        <- vector("list",nc)
    names(hlist) <- lcf
}

# Run through the cells:
for(xl in lcf) {
    ind   <- (data[["cf"]]==xl)
    Xxl   <- X[ind,,drop=FALSE]
    nind  <- nrow(Xxl)
    fyxl  <- fy[[xl]]
    yxl   <- matrix(y[ind],ncol=K,byrow=TRUE)[,1] # All columns are identical.
    ymiss <- is.na(yxl)
    ny    <- length(yxl)
    xxx   <- .Fortran(
        "gethgl",
        NAOK=TRUE,
        nd=as.integer(nd),
        ndistr=as.integer(ndistr),
        fy=as.double(fyxl),
        gmu=as.double(gmu[ind]),
        sd=as.double(sd[ind]),
        lambda=as.double(lambda[ind]),
        p=as.double(p[ind]),
        ashp=as.double(ashp[ind]),
        bshp=as.double(bshp[ind]),
        phimat=as.double(Rho),
        y=as.double(yxl),
        ymiss=as.integer(ymiss),
        tdm=as.double(t(Xxl)),
        nind=as.integer(nind),
        n=as.integer(ny),
        tpm=as.double(tpm),
        xispd=as.double(ispd),
        size=as.integer(size),
        nbot=as.integer(nbot),
        ntop=as.integer(ntop),
        d1pi=as.double(d1pi),
        d2pi=as.double(d2pi),
        kstate=as.integer(K),
        npar=as.integer(npar),
        npt=as.integer(npt),
        nyv=as.integer(nyv),
        nxc=as.integer(nxc),
        d1p=as.double(d1p),
        d2p=as.double(d2p),
        d1f=double(K*npar),
        d2f=double(K*npar*npar),
        d1a=double(K),
        d1b=double(K),
        d2aa=double(K),
        d2ab=double(K),
        d2bb=double(K),
        alpha=double(K),
        alphw=double(K),
        a=double(K*npar),
        b=double(K*npar*npar),
        aw=double(K*npar),
        bw=double(K*npar*npar),
        xlc=double(ny),
        hess=double(npar*npar),
        PACKAGE="eglhmm"
        )
    xlc <- xxx$xlc
    if(nd >= 1) alist[[xl]]  <- matrix(xxx$a,K,npar)/xlc[ny]
    xll[xl] <- sum(log(xlc))
    if(nd == 2) hlist[[xl]]  <- matrix(xxx$hess,npar,npar)
} 
ll <- sum(xll)
if(nd == 0) return(list(ll=ll))
a           <- array(unlist(alist),c(K,npar,nc))
grad        <- apply(a,2,sum)
names(grad) <- nms
if(nd == 1) return(list(ll=ll,gradient=grad))
hpre           <- array(unlist(hlist),c(npar,npar,nc))
Hessian        <- apply(hpre,c(1,2),sum)
dimnames(Hessian) <- list(nms,nms)
list(ll=ll,gradient=grad,Hessian=Hessian)
}
